yew-macro/src/props/prop.rs

```diff
@@ -0,0 +1,245 @@
+        self.position(key).and_then(|i| self.0.get(i))
+    }
+
+    /// Pop the first prop with the given key.
+    pub fn pop(&mut self, key: &str) -> Option<Prop> {
+        self.position(key).map(|i| self.0.remove(i))
+    }
+
+    /// Pop the prop with the given key and error if there are multiple ones.
+    pub fn pop_unique(&mut self, key: &str) -> syn::Result<Option<Prop>> {
+        let prop = self.pop(key);
+        if prop.is_some() {
+            if let Some(other_prop) = self.pop(key) {
+                return Err(syn::Error::new_spanned(
+                    other_prop.label,
+                    format!("`{}` can only be specified once", key),
+                ));
+            }
+        }
+
+        Ok(prop)
+    }
+    /// Pop the prop with the given key and error if it uses the optional attribute syntax.
+    pub fn pop_nonoptional(&mut self, key: &str) -> syn::Result<Option<Prop>> {
+        match self.pop_unique(key) {
+            Ok(Some(prop)) => {
+                prop.ensure_not_optional()?;
+                Ok(Some(prop))
+            }
+            res => res,
+        }
+    }
+
+    /// Turn the props into a vector of `Prop`.
+    pub fn into_vec(self) -> Vec<Prop> {
+        self.0
+    }
+
+    /// Iterate over all duplicate props in order of appearance.
+    fn iter_duplicates(&self) -> impl Iterator<Item = &Prop> {
+        self.0.windows(2).filter_map(|pair| {
+            let (a, b) = (&pair[0], &pair[1]);
+
+            if a.label == b.label {
+                Some(b)
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Remove and return all props for which `filter` returns `true`.
+    pub fn drain_filter(&mut self, filter: impl FnMut(&Prop) -> bool) -> PropList {
+        let (drained, others) = self.0.drain(..).partition(filter);
+        self.0 = others;
+        PropList(drained)
+    }
+
+    /// Run the given function for all props and aggregate the errors.
+    /// If there's at least one error, the result will be `Result::Err`.
+    pub fn check_all(&self, f: impl FnMut(&Prop) -> syn::Result<()>) -> syn::Result<()> {
+        crate::join_errors(self.0.iter().map(f).filter_map(Result::err))
+    }
+
+    /// Return an error for all duplicate props.
+    pub fn check_no_duplicates(&self) -> syn::Result<()> {
+        crate::join_errors(self.iter_duplicates().map(|prop| {
+            syn::Error::new_spanned(
+                &prop.label,
+                format!(
+                    "`{}` can only be specified once but is given here again",
+                    prop.label
+                ),
+            )
+        }))
+    }
+}
+impl Parse for PropList {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut props: Vec<Prop> = Vec::new();
+        while !input.is_empty() {
+            props.push(input.parse()?);
+        }
+
+        props.sort_by(|a, b| Self::cmp_label(&a.label.to_string(), &b.label.to_string()));
```

Comment:
I was surprised that prop parsing will also sort but this is pretty nice! Maybe we could rename `PropList` to something like `SortedPropList` and have `SortedPropList::new(..)` do the sorting or something? I would just like it to be more explicit :)